File: rust/src/credential.rs
use super::{ffi::c_void, utils::SeekableReader};
use anyhow::Result as AnyResult;
use qiniu_sdk::http::{HeaderMap, HeaderName, HeaderValue, Method, Uri};
use std::{io::Result as IoResult, time::Duration};

#[derive(Clone, Debug)]
pub struct Credential(qiniu_sdk::credential::Credential);

pub fn new_credential(access_key: &str, secret_key: &str) -> Box<Credential> {
    Box::new(Credential(qiniu_sdk::credential::Credential::new(
        access_key, secret_key,
    )))
}

pub fn credential_clone(credential: &Credential) -> Box<Credential> {
    Box::new(credential.to_owned())
}

pub fn credential_get_access_key(credential: &Credential) -> &str {
    credential.0.access_key().as_ref()
}

pub fn credential_get_secret_key(credential: &Credential) -> &str {
    credential.0.secret_key().as_ref()
}

pub fn credential_sign(credential: &Credential, data: &[u8]) -> String {
    credential.0.sign(data)
}

pub fn credential_sign_with_data(credential: &Credential, data: &[u8]) -> String {
    credential.0.sign_with_data(data)
}

pub fn credential_sign_reader(credential: &Credential, stream: *mut c_void) -> IoResult<String> {
    credential.0.sign_reader(&mut SeekableReader::new(stream))
}

pub fn credential_authorization_v1_for_request(
    credential: &Credential,
    url: &str,
    content_type: &str,
    body: &[u8],
) -> AnyResult<String> {
    let url = parse_uri(url)?;
    let content_type = parse_content_type(content_type)?;
    Ok(credential
        .0
        .authorization_v1_for_request(&url, content_type.as_ref(), body))
}

pub fn credential_authorization_v1_for_request_with_body_reader(
    credential: &Credential,
    url: &str,
    content_type: &str,
    stream: *mut c_void,
) -> AnyResult<String> {
    let url = parse_uri(url)?;
    let content_type = parse_content_type(content_type)?;
    let authorization = credential.0.authorization_v1_for_request_with_body_reader(
        &url,
        content_type.as_ref(),
        &mut SeekableReader::new(stream),
    )?;
    Ok(authorization)
}

pub fn credential_authorization_v2_for_request(
    credential: &Credential,
    method: &str,
    url: &str,
    headers: &[[&str; 2]],
    body: &[u8],
) -> AnyResult<String> {
    let method = parse_method(method)?;
    let url = parse_uri(url)?;
    let headers = parse_header_map(headers)?;
    Ok(credential
        .0
        .authorization_v2_for_request(&method, &url, &headers, body))
}

pub fn credential_authorization_v2_for_request_with_body_reader(
    credential: &Credential,
    method: &str,
    url: &str,
    headers: &[[&str; 2]],
    stream: *mut c_void,
) -> AnyResult<String> {
    let method = parse_method(method)?;
    let url = parse_uri(url)?;
    let headers = parse_header_map(headers)?;
    let authorization = credential.0.authorization_v2_for_request_with_body_reader(
        &method,
        &url,
        &headers,
        &mut SeekableReader::new(stream),
    )?;
    Ok(authorization)
}

pub fn credential_sign_download_url(
    credential: &Credential,
    url: &str,
    duration: u64,
) -> AnyResult<String> {
    let url = parse_uri(url)?;
    let duration = Duration::from_nanos(duration);
    Ok(credential.0.sign_download_url(url, duration).to_string())
}

#[derive(Clone, Debug)]
pub struct CredentialProvider(Box<dyn qiniu_sdk::credential::CredentialProvider>);

impl qiniu_sdk::credential::CredentialProvider for CredentialProvider {
    fn get(
        &self,
        opts: qiniu_sdk::credential::GetOptions,
    ) -> IoResult<qiniu_sdk::credential::GotCredential> {
        self.0.get(opts)
    }
}

#[derive(Clone, Debug)]
pub struct GetCredentialOptions(qiniu_sdk::credential::GetOptions);

#[derive(Clone, Debug)]
pub struct GotCredential(qiniu_sdk::credential::GotCredential);

pub fn credential_provider_get(
    provider: &CredentialProvider,
    options: &GetCredentialOptions,
) -> IoResult<Box<GotCredential>> {
    provider
        .0
        .get(options.0)
        .map(|r| Box::new(GotCredential(r)))
}

pub fn new_static_credential_provider(credential: Box<Credential>) -> Box<CredentialProvider> {
    Box::new(CredentialProvider(Box::new(credential.0)))
}

pub fn credential_provider_clone(provider: &CredentialProvider) -> Box<CredentialProvider> {
    Box::new(provider.to_owned())
}

pub fn got_credential_get_credential(got_credential: &GotCredential) -> Box<Credential> {
    Box::new(Credential(got_credential.0.credential().to_owned()))
}

pub fn new_get_credential_options() -> Box<GetCredentialOptions> {
    Box::new(GetCredentialOptions(
        qiniu_sdk::credential::GetOptions::default(),
    ))
}

pub fn get_credential_options_clone(options: &GetCredentialOptions) -> Box<GetCredentialOptions> {
    Box::new(options.to_owned())
}

pub fn new_global_credential_provider() -> Box<CredentialProvider> {
    Box::new(CredentialProvider(Box::new(
        qiniu_sdk::credential::GlobalCredentialProvider,
    )))
}

pub fn global_credential_provider_setup(credential: &Credential) {
    qiniu_sdk::credential::GlobalCredentialProvider::setup(credential.0.to_owned())
}

pub fn global_credential_provider_clear() {
    qiniu_sdk::credential::GlobalCredentialProvider::clear()
}

pub fn new_env_credential_provider() -> Box<CredentialProvider> {
    Box::new(CredentialProvider(Box::new(
        qiniu_sdk::credential::EnvCredentialProvider,
    )))
}

pub fn env_credential_provider_setup(credential: &Credential) {
    qiniu_sdk::credential::EnvCredentialProvider::setup(&credential.0)
}

pub fn env_credential_provider_clear() {
    qiniu_sdk::credential::EnvCredentialProvider::clear()
}

#[derive(Clone, Debug)]
pub struct ChainCredentialsProviderBuilder(
    Box<qiniu_sdk::credential::ChainCredentialsProviderBuilder>,
);

pub fn new_chain_credentials_provider_builder(
    provider: Box<CredentialProvider>,
) -> Box<ChainCredentialsProviderBuilder> {
    Box::new(ChainCredentialsProviderBuilder(Box::new(
        qiniu_sdk::credential::ChainCredentialsProviderBuilder::new(provider.0),
    )))
}

pub fn chain_credentials_provider_builder_append_credential(
    builder: &mut ChainCredentialsProviderBuilder,
    provider: Box<CredentialProvider>,
) {
    builder.0.append_credential(provider.0);
}

pub fn chain_credentials_provider_builder_prepend_credential(
    builder: &mut ChainCredentialsProviderBuilder,
    provider: Box<CredentialProvider>,
) {
    builder.0.prepend_credential(provider.0);
}

pub fn chain_credentials_provider_builder_build(
    builder: &mut ChainCredentialsProviderBuilder,
) -> Box<CredentialProvider> {
    Box::new(CredentialProvider(Box::new(builder.0.build())))
}

fn parse_uri(url: &str) -> AnyResult<Uri> {
    let url = url.parse::<Uri>()?;
    Ok(url)
}

fn parse_method(method: &str) -> AnyResult<Method> {
    let method = method.parse::<Method>()?;
    Ok(method)
}

fn parse_content_type(content_type: &str) -> AnyResult<Option<HeaderValue>> {
    let content_type = if content_type.is_empty() {
        None
    } else {
        Some(content_type.parse::<HeaderValue>()?)
    };
    Ok(content_type)
}

fn parse_header_name(header_name: &str) -> AnyResult<HeaderName> {
    let header_name = header_name.parse::<HeaderName>()?;
    Ok(header_name)
}

fn parse_header_value(header_value: &str) -> AnyResult<HeaderValue> {
    let header_value = header_value.parse::<HeaderValue>()?;
    Ok(header_value)
}

fn parse_header_map(headers: &[[&str; 2]]) -> AnyResult<HeaderMap> {
    headers
        .into_iter()
        .map(|[header_name, header_value]| {
            let header_name = parse_header_name(header_name)?;
            let header_value = parse_header_value(header_value)?;
            Ok((header_name, header_value))
        })
        .collect()
}

File: cpp/credential.cpp
#include "includes/qiniu_sdk.h"
#include <type_traits>
#include <array>
#include <map>
#include <vector>

namespace qiniu_sdk
{
    namespace credential
    {
        std::vector<std::array<rust::Str, 2>> convert_headers(const std::map<std::string, std::string> &) noexcept;

        Credential::Credential(const std::string &access_key, const std::string &secret_key) noexcept : inner(_internal::rust_sdk_ffi::new_credential(rust::Str(access_key), rust::Str(secret_key)))
        {
        }

        Credential::Credential(const Credential &credential) noexcept : inner(_internal::rust_sdk_ffi::credential_clone(*credential))
        {
        }

        Credential::Credential(const GotCredential &credential) noexcept : inner(_internal::rust_sdk_ffi::got_credential_get_credential(*credential))
        {
        }

        Credential::Credential(rust::Box<_internal::rust_sdk_ffi::Credential> &&credential) noexcept : inner(std::move(credential))
        {
        }

        const _internal::rust_sdk_ffi::Credential &Credential::operator*() const noexcept
        {
            return *this->inner;
        }

        _internal::rust_sdk_ffi::Credential &Credential::operator*() noexcept
        {
            return *this->inner;
        }

        std::string Credential::get_access_key() const noexcept
        {
            return std::string(_internal::rust_sdk_ffi::credential_get_access_key(*this->inner));
        }

        std::string Credential::get_secret_key() const noexcept
        {
            return std::string(_internal::rust_sdk_ffi::credential_get_secret_key(*this->inner));
        }

        std::string Credential::sign(const uint8_t *data, size_t len) const
        {
            auto data_slice = rust::Slice<const uint8_t>(data, len);
            return std::string(_internal::rust_sdk_ffi::credential_sign(*this->inner, data_slice));
        }

        std::string Credential::sign_with_data(const uint8_t *data, size_t len) const
        {
            auto data_slice = rust::Slice<const uint8_t>(data, len);
            return std::string(_internal::rust_sdk_ffi::credential_sign_with_data(*this->inner, data_slice));
        }

        std::string Credential::sign_reader(std::istream *stream) const
        {
            return std::string(_internal::rust_sdk_ffi::credential_sign_reader(*this->inner, static_cast<void *>(stream)));
        }

        std::string Credential::authorization_v1_for_request(const std::string &url, const std::string &content_type, const uint8_t *body, size_t body_size) const
        {
            auto url_str = rust::Str(url);
            auto content_type_str = rust::Str(content_type);
            auto body_slice = rust::Slice<const uint8_t>(body, body_size);
            return std::string(_internal::rust_sdk_ffi::credential_authorization_v1_for_request(*this->inner, url_str, content_type_str, body_slice));
        }

        std::string Credential::authorization_v1_for_request_with_body_reader(const std::string &url, const std::string &content_type, std::istream *stream) const
        {
            auto url_str = rust::Str(url);
            auto content_type_str = rust::Str(content_type);
            return std::string(_internal::rust_sdk_ffi::credential_authorization_v1_for_request_with_body_reader(*this->inner, url_str, content_type_str, static_cast<void *>(stream)));
        }

        std::string Credential::authorization_v2_for_request(const std::string &method, const std::string &url, const std::map<std::string, std::string> &headers, const uint8_t *body, size_t body_size) const
        {
            auto url_str = rust::Str(url);
            auto method_str = rust::Str(method);
            auto header_vec = convert_headers(headers);
            auto header_slice = rust::Slice<const ::std::array<rust::Str, 2>>(&header_vec[0], header_vec.size());
            auto body_slice = rust::Slice<const uint8_t>(body, body_size);
            return std::string(_internal::rust_sdk_ffi::credential_authorization_v2_for_request(*this->inner, method_str, url_str, header_slice, body_slice));
        }

        std::string Credential::authorization_v2_for_request_with_body_reader(const std::string &method, const std::string &url, const std::map<std::string, std::string> &headers, std::istream *stream) const
        {
            auto url_str = rust::Str(url);
            auto method_str = rust::Str(method);
            auto header_vec = convert_headers(headers);
            auto header_slice = rust::Slice<const ::std::array<rust::Str, 2>>(&header_vec[0], header_vec.size());
            return std::string(_internal::rust_sdk_ffi::credential_authorization_v2_for_request_with_body_reader(*this->inner, method_str, url_str, header_slice, static_cast<void *>(stream)));
        }

        std::string Credential::sign_download_url(const std::string &url, const std::chrono::nanoseconds &lifetime) const
        {
            auto url_str = rust::Str(url);
            return std::string(_internal::rust_sdk_ffi::credential_sign_download_url(*this->inner, url_str, lifetime.count()));
        }

        GotCredential::GotCredential(rust::Box<_internal::rust_sdk_ffi::GotCredential> &&got_credential) noexcept : inner(std::move(got_credential))
        {
        }

        const _internal::rust_sdk_ffi::GotCredential &GotCredential::operator*() const noexcept
        {
            return *this->inner;
        }

        _internal::rust_sdk_ffi::GotCredential &GotCredential::operator*() noexcept
        {
            return *this->inner;
        }

        GetCredentialOptions::GetCredentialOptions() noexcept : inner(_internal::rust_sdk_ffi::new_get_credential_options())
        {
        }

        GetCredentialOptions::GetCredentialOptions(const GetCredentialOptions &options) noexcept : inner(_internal::rust_sdk_ffi::get_credential_options_clone(*options.inner))
        {
        }

        GetCredentialOptions::GetCredentialOptions(rust::Box<_internal::rust_sdk_ffi::GetCredentialOptions> &&options) noexcept : inner(std::move(options))
        {
        }

        const _internal::rust_sdk_ffi::GetCredentialOptions &GetCredentialOptions::operator*() const noexcept
        {
            return *this->inner;
        }

        _internal::rust_sdk_ffi::GetCredentialOptions &GetCredentialOptions::operator*() noexcept
        {
            return *this->inner;
        }

        CredentialProvider::CredentialProvider(Credential credential) noexcept : inner(_internal::rust_sdk_ffi::new_static_credential_provider(std::move(credential.inner)))
        {
        }

        CredentialProvider::CredentialProvider(const CredentialProvider &provider) noexcept : inner(_internal::rust_sdk_ffi::credential_provider_clone(*provider.inner))
        {
        }

        CredentialProvider::CredentialProvider(rust::Box<_internal::rust_sdk_ffi::CredentialProvider> &&provider) noexcept : inner(std::move(provider))
        {
        }

        const _internal::rust_sdk_ffi::CredentialProvider &CredentialProvider::operator*() const noexcept
        {
            return *this->inner;
        }

        _internal::rust_sdk_ffi::CredentialProvider &CredentialProvider::operator*() noexcept
        {
            return *this->inner;
        }

        GotCredential CredentialProvider::get(const GetCredentialOptions &options) const
        {
            return _internal::rust_sdk_ffi::credential_provider_get(*this->inner, *options);
        }

        GlobalCredentialProvider::GlobalCredentialProvider() noexcept : CredentialProvider(_internal::rust_sdk_ffi::new_global_credential_provider())
        {
        }

        void GlobalCredentialProvider::setup(const Credential &credential) noexcept
        {
            _internal::rust_sdk_ffi::global_credential_provider_setup(*credential);
        }

        void GlobalCredentialProvider::clear() noexcept
        {
            _internal::rust_sdk_ffi::global_credential_provider_clear();
        }

        EnvCredentialProvider::EnvCredentialProvider() noexcept : CredentialProvider(_internal::rust_sdk_ffi::new_env_credential_provider())
        {
        }

        void EnvCredentialProvider::setup(const Credential &credential) noexcept
        {
            _internal::rust_sdk_ffi::env_credential_provider_setup(*credential);
        }

        void EnvCredentialProvider::clear() noexcept
        {
            _internal::rust_sdk_ffi::env_credential_provider_clear();
        }

        ChainCredentialsProviderBuilder::ChainCredentialsProviderBuilder(CredentialProvider provider) noexcept : inner(_internal::rust_sdk_ffi::new_chain_credentials_provider_builder(std::move(provider.inner)))
        {
        }

        ChainCredentialsProviderBuilder::ChainCredentialsProviderBuilder(rust::Box<_internal::rust_sdk_ffi::ChainCredentialsProviderBuilder> &&provider) noexcept : inner(std::move(provider))
        {
        }

        const _internal::rust_sdk_ffi::ChainCredentialsProviderBuilder &ChainCredentialsProviderBuilder::operator*() const noexcept
        {
            return *this->inner;
        }

        _internal::rust_sdk_ffi::ChainCredentialsProviderBuilder &ChainCredentialsProviderBuilder::operator*() noexcept
        {
            return *this->inner;
        }

        void ChainCredentialsProviderBuilder::append_credential(CredentialProvider provider) noexcept
        {
            _internal::rust_sdk_ffi::chain_credentials_provider_builder_append_credential(*this->inner, std::move(provider.inner));
        }

        void ChainCredentialsProviderBuilder::prepend_credential(CredentialProvider provider) noexcept
        {
            _internal::rust_sdk_ffi::chain_credentials_provider_builder_prepend_credential(*this->inner, std::move(provider.inner));
        }

        ChainCredentialsProvider ChainCredentialsProviderBuilder::build() noexcept
        {
            return ChainCredentialsProvider(*this);
        }

        ChainCredentialsProvider::ChainCredentialsProvider(ChainCredentialsProviderBuilder &provider) noexcept : CredentialProvider(_internal::rust_sdk_ffi::chain_credentials_provider_builder_build(*provider))
        {
        }
    }
}
